File: Data Structures & Algorithms/two-integer-sum-ii/submission-4.rs
impl Solution {
    pub fn two_sum(numbers: Vec<i32>, target: i32) -> Vec<i32> {
        // O(1) space
        // O(N)
        // Sorted numbers
        // Return only 2 numbers to equal to target
        // Only 1 exact Solution
        let mut right = numbers.len() - 1;
        let mut left = 0;
        while left < right {
            if numbers[left] + numbers[right] == target {
                return vec![(left + 1) as i32, (right + 1) as i32]
            } else if numbers[left] + numbers[right] > target {
                right -= 1;
            } else {
                left += 1;
            }
        }
        unreachable!()
    }
}

File: Data Structures & Algorithms/valid-palindrome-ii/submission-4.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        def is_palindrome(left: int, right: int) -> bool:
            while left < right:
                if s[left] != s[right]:
                    return False
                left += 1
                right -= 1
            return True

        l, r = 0, len(s) - 1
        while l < r:
            if s[l] != s[r]:
                # Try deleting the left character OR the right character
                return is_palindrome(l + 1, r) or is_palindrome(l, r - 1)
            l += 1
            r -= 1
        return True